File: validation/make_synthetic_statcast.py
"""
Generate a synthetic, Statcast-shaped pitch-level dataset.

This exists ONLY so the Python modeling chunks in mlb_statcast.qmd can be
executed and checked end-to-end in environments with no Baseball Savant
access. Column names, dtypes, units and missingness patterns mirror
`baseballr::scrape_statcast_savant(player_type = "pitcher")`; the numbers
are simulated and carry no baseball meaning.
"""

import numpy as np
import pandas as pd

RNG = np.random.default_rng(20260823)

# (pitch_name, mean velo, velo sd, mean pfx_x arm-side ft, mean pfx_z ft, mean spin)
PITCH_TYPES = [
    ("4-Seam Fastball", 94.0, 2.1, 0.62, 1.40, 2300, 0.235),
    ("Sinker",          93.0, 2.0, 1.30, 0.75, 2150, 0.155),
    ("Cutter",          89.5, 2.0, -0.15, 0.75, 2400, 0.245),
    ("Slider",          85.0, 2.6, -0.75, 0.10, 2450, 0.355),
    ("Sweeper",         81.5, 2.4, -1.35, 0.05, 2600, 0.365),
    ("Curveball",       79.0, 2.7, -0.90, -0.65, 2600, 0.330),
    ("Knuckle Curve",   80.5, 2.5, -0.80, -0.50, 2650, 0.325),
    ("Slurve",          81.0, 2.5, -1.10, -0.30, 2550, 0.330),
    ("Split-Finger",    86.5, 2.2, 0.70, 0.10, 1500, 0.395),
    ("Changeup",        85.0, 2.3, 1.15, 0.35, 1800, 0.310),
    ("Forkball",        83.0, 2.0, 0.60, 0.05, 1300, 0.330),
    ("Eephus",          58.0, 4.0, 0.30, -1.20, 1400, 0.200),
    ("Knuckleball",     70.0, 3.0, 0.10, 0.10,  900, 0.260),
    ("Slow Curve",      68.0, 3.0, -0.70, -1.10, 2200, 0.280),
]

N_PITCHERS = 300


def _pitcher_frame():
    ids = np.arange(1, N_PITCHERS + 1)
    return pd.DataFrame({
        "pitcher": 600000 + ids,
        "player_name": [f"Sim{i:03d}, Pitcher" for i in ids],
        "p_throws": RNG.choice(["R", "L"], N_PITCHERS, p=[0.72, 0.28]),
        # latent per-pitcher talent, shifts both whiff propensity and run value
        "talent": RNG.normal(0, 0.30, N_PITCHERS),
        "velo_shift": RNG.normal(0, 1.7, N_PITCHERS),
        "ext_base": RNG.normal(6.4, 0.32, N_PITCHERS),
        "arm_base": RNG.normal(42, 13, N_PITCHERS).clip(0, 90),
    })


def simulate(n_pitch_rows=140_000):
    pit = _pitcher_frame()
    pidx = RNG.integers(0, N_PITCHERS, n_pitch_rows)
    tidx = RNG.integers(0, len(PITCH_TYPES), n_pitch_rows)
    # make arsenal realistic: fastballs far more common than Eephus
    weights = np.array([0.30, 0.14, 0.07, 0.15, 0.07, 0.07, 0.03,
                        0.01, 0.03, 0.11, 0.005, 0.002, 0.003, 0.005])
    tidx = RNG.choice(len(PITCH_TYPES), n_pitch_rows, p=weights / weights.sum())

    names = np.array([p[0] for p in PITCH_TYPES])
    mv = np.array([p[1] for p in PITCH_TYPES])
    sv = np.array([p[2] for p in PITCH_TYPES])
    mx = np.array([p[3] for p in PITCH_TYPES])
    mz = np.array([p[4] for p in PITCH_TYPES])
    ms = np.array([p[5] for p in PITCH_TYPES])
    mw = np.array([p[6] for p in PITCH_TYPES])

    throws = pit["p_throws"].to_numpy()[pidx]
    sign = np.where(throws == "R", 1.0, -1.0)   # arm-side -> raw pfx_x sign

    velo = RNG.normal(mv[tidx] + pit["velo_shift"].to_numpy()[pidx], sv[tidx])
    ext = RNG.normal(pit["ext_base"].to_numpy()[pidx], 0.18)
    arm = RNG.normal(pit["arm_base"].to_numpy()[pidx], 4.0).clip(-10, 95)
    spin = RNG.normal(ms[tidx] + 60 * pit["talent"].to_numpy()[pidx], 180).clip(300, 3600)

    pfx_x_arm = RNG.normal(mx[tidx], 0.30)      # arm-side convention
    pfx_z = RNG.normal(mz[tidx], 0.30)
    rel_x_arm = RNG.normal(1.9 - 0.014 * arm, 0.35)
    rel_z = RNG.normal(4.4 + 0.012 * arm, 0.35)

    # ---- latent pitch quality: deliberately NON-linear in velo and pfx_z ----
    z_velo = (velo - mv[tidx]) / sv[tidx]
    lin = (
        0.30 * z_velo
        + 0.22 * (spin - ms[tidx]) / 200
        + 0.18 * (ext - 6.4) / 0.3
        + 0.20 * np.abs(pfx_x_arm - mx[tidx])
        + pit["talent"].to_numpy()[pidx]
    )
    nonlin = 0.14 * z_velo**2 - 0.16 * (pfx_z - mz[tidx]) ** 2
    quality = lin + nonlin

    p_swing = 1 / (1 + np.exp(-(0.05 + 0.25 * z_velo)))
    p_swing = np.clip(p_swing * 0.94, 0.05, 0.95)
    swing = RNG.random(n_pitch_rows) < p_swing

    logit_whiff = np.log(mw[tidx] / (1 - mw[tidx])) + 0.85 * quality
    whiff = swing & (RNG.random(n_pitch_rows) < 1 / (1 + np.exp(-logit_whiff)))

    desc = np.where(
        whiff, "swinging_strike",
        np.where(swing,
                 RNG.choice(["foul", "hit_into_play", "foul_tip"], n_pitch_rows, p=[0.55, 0.40, 0.05]),
                 RNG.choice(["ball", "called_strike", "blocked_ball"], n_pitch_rows, p=[0.55, 0.40, 0.05])),
    )

    # run value: better stuff -> more negative delta_run_exp for the pitcher
    delta_run_exp = RNG.normal(-0.010 * quality, 0.115)
    delta_run_exp = np.where(desc == "hit_into_play",
                             delta_run_exp + RNG.normal(0.10, 0.45, n_pitch_rows),
                             delta_run_exp)

    df = pd.DataFrame({
        "player_name": pit["player_name"].to_numpy()[pidx],
        "pitcher": pit["pitcher"].to_numpy()[pidx],
        "p_throws": throws,
        "stand": RNG.choice(["R", "L"], n_pitch_rows, p=[0.55, 0.45]),
        "pitch_name": names[tidx],
        "release_speed": velo.round(1),
        "release_pos_x": (sign * rel_x_arm).round(2),
        "release_pos_z": rel_z.round(2),
        "release_spin_rate": spin.round(0),
        "release_extension": ext.round(2),
        "spin_axis": RNG.integers(0, 360, n_pitch_rows),
        "arm_angle": arm.round(1),
        "pfx_x": (sign * pfx_x_arm).round(2),
        "pfx_z": pfx_z.round(2),
        "plate_x": RNG.normal(0, 0.85, n_pitch_rows).round(2),
        "plate_z": RNG.normal(2.4, 0.75, n_pitch_rows).round(2),
        "description": desc,
        "delta_run_exp": delta_run_exp.round(4),
    })

    # Statcast missingness: arm_angle and spin are absent for a slice of rows,
    # and a handful of pitchers are missing arm_angle entirely -- this is what
    # produced the "'data' contains 5 incomplete rows" failure.
    df.loc[RNG.random(len(df)) < 0.02, "release_spin_rate"] = np.nan
    df.loc[RNG.random(len(df)) < 0.03, "arm_angle"] = np.nan
    dead = RNG.choice(pit["player_name"], 5, replace=False)
    df.loc[df["player_name"].isin(dead), "arm_angle"] = np.nan
    df.loc[RNG.random(len(df)) < 0.005, "release_extension"] = np.nan
    return df


if __name__ == "__main__":
    out = simulate()
    out.to_csv("synthetic_statcast.csv", index=False)
    print(out.shape)
    print(out["pitch_name"].value_counts())

File: validation/make_synthetic_statcast.R
# Generate a synthetic, Statcast-shaped pitch-level dataset.
#
# This exists ONLY so the modelling chunks in mlb_statcast.qmd can be executed
# and checked end-to-end in environments with no Baseball Savant access. Column
# names, units and missingness patterns mirror
# `baseballr::scrape_statcast_savant(player_type = "pitcher")`; the numbers are
# simulated and carry no baseball meaning.
suppressMessages({library(dplyr); library(readr)})
set.seed(20260823)

PITCH_TYPES <- tribble(
  ~pitch_name,        ~mv,  ~sv, ~mx,   ~mz,   ~ms,  ~mw,   ~wt,
  "4-Seam Fastball", 94.0, 2.1,  0.62,  1.40, 2300, 0.235, 0.300,
  "Sinker",          93.0, 2.0,  1.30,  0.75, 2150, 0.155, 0.140,
  "Cutter",          89.5, 2.0, -0.15,  0.75, 2400, 0.245, 0.070,
  "Slider",          85.0, 2.6, -0.75,  0.10, 2450, 0.355, 0.150,
  "Sweeper",         81.5, 2.4, -1.35,  0.05, 2600, 0.365, 0.070,
  "Curveball",       79.0, 2.7, -0.90, -0.65, 2600, 0.330, 0.070,
  "Knuckle Curve",   80.5, 2.5, -0.80, -0.50, 2650, 0.325, 0.030,
  "Slurve",          81.0, 2.5, -1.10, -0.30, 2550, 0.330, 0.010,
  "Split-Finger",    86.5, 2.2,  0.70,  0.10, 1500, 0.395, 0.030,
  "Changeup",        85.0, 2.3,  1.15,  0.35, 1800, 0.310, 0.110,
  "Forkball",        83.0, 2.0,  0.60,  0.05, 1300, 0.330, 0.005,
  "Eephus",          58.0, 4.0,  0.30, -1.20, 1400, 0.200, 0.002,
  "Knuckleball",     70.0, 3.0,  0.10,  0.10,  900, 0.260, 0.003,
  "Slow Curve",      68.0, 3.0, -0.70, -1.10, 2200, 0.280, 0.005
)

N_PITCHERS <- 300L
N_ROWS     <- 140000L

pit <- tibble(
  id          = seq_len(N_PITCHERS),
  pitcher     = 600000L + seq_len(N_PITCHERS),
  player_name = sprintf("Sim%03d, Pitcher", seq_len(N_PITCHERS)),
  p_throws    = sample(c("R", "L"), N_PITCHERS, TRUE, c(0.72, 0.28)),
  talent      = rnorm(N_PITCHERS, 0, 0.30),
  velo_shift  = rnorm(N_PITCHERS, 0, 1.7),
  ext_base    = rnorm(N_PITCHERS, 6.4, 0.32),
  arm_base    = pmin(pmax(rnorm(N_PITCHERS, 42, 13), 0), 90)
)

pidx <- sample.int(N_PITCHERS, N_ROWS, replace = TRUE)
tidx <- sample.int(nrow(PITCH_TYPES), N_ROWS, replace = TRUE,
                   prob = PITCH_TYPES$wt / sum(PITCH_TYPES$wt))

throws <- pit$p_throws[pidx]
sgn    <- if_else(throws == "R", 1, -1)   # arm-side -> raw pfx_x sign

mv <- PITCH_TYPES$mv[tidx]; sv <- PITCH_TYPES$sv[tidx]
mx <- PITCH_TYPES$mx[tidx]; mz <- PITCH_TYPES$mz[tidx]
ms <- PITCH_TYPES$ms[tidx]; mw <- PITCH_TYPES$mw[tidx]
talent <- pit$talent[pidx]

velo <- rnorm(N_ROWS, mv + pit$velo_shift[pidx], sv)
ext  <- rnorm(N_ROWS, pit$ext_base[pidx], 0.18)
arm  <- pmin(pmax(rnorm(N_ROWS, pit$arm_base[pidx], 4.0), -10), 95)
spin <- pmin(pmax(rnorm(N_ROWS, ms + 60 * talent, 180), 300), 3600)

pfx_x_arm <- rnorm(N_ROWS, mx, 0.30)      # arm-side convention
pfx_z     <- rnorm(N_ROWS, mz, 0.30)
rel_x_arm <- rnorm(N_ROWS, 1.9 - 0.014 * arm, 0.35)
rel_z     <- rnorm(N_ROWS, 4.4 + 0.012 * arm, 0.35)

# Latent pitch quality: deliberately NON-linear in velocity and vertical break,
# so the linearity tests have something real to find.
z_velo <- (velo - mv) / sv
quality <- 0.30 * z_velo + 0.22 * (spin - ms) / 200 + 0.18 * (ext - 6.4) / 0.3 +
  0.20 * abs(pfx_x_arm - mx) + talent +
  0.14 * z_velo^2 - 0.16 * (pfx_z - mz)^2

p_swing <- pmin(pmax(plogis(0.05 + 0.25 * z_velo) * 0.94, 0.05), 0.95)
swing   <- runif(N_ROWS) < p_swing
whiff   <- swing & (runif(N_ROWS) < plogis(qlogis(mw) + 0.85 * quality))

desc <- if_else(
  whiff, "swinging_strike",
  if_else(swing,
          sample(c("foul", "hit_into_play", "foul_tip"), N_ROWS, TRUE, c(.55, .40, .05)),
          sample(c("ball", "called_strike", "blocked_ball"), N_ROWS, TRUE, c(.55, .40, .05)))
)

dre <- rnorm(N_ROWS, -0.010 * quality, 0.115)
dre <- if_else(desc == "hit_into_play", dre + rnorm(N_ROWS, 0.10, 0.45), dre)

game_pk <- rep_len(seq_len(ceiling(N_ROWS / 300)), N_ROWS)
at_bat  <- rep_len(seq_len(80), N_ROWS)

out <- tibble(
  game_pk, at_bat_number = at_bat,
  pitch_number = ave(seq_len(N_ROWS), paste(game_pk, at_bat), FUN = seq_along),
  player_name = pit$player_name[pidx], pitcher = pit$pitcher[pidx],
  p_throws = throws, stand = sample(c("R", "L"), N_ROWS, TRUE, c(.55, .45)),
  pitch_name = PITCH_TYPES$pitch_name[tidx],
  release_speed = round(velo, 1),
  release_pos_x = round(sgn * rel_x_arm, 2),
  release_pos_z = round(rel_z, 2),
  release_spin_rate = round(spin),
  release_extension = round(ext, 2),
  spin_axis = sample.int(360, N_ROWS, TRUE) - 1L,
  arm_angle = round(arm, 1),
  pfx_x = round(sgn * pfx_x_arm, 2),
  pfx_z = round(pfx_z, 2),
  plate_x = round(rnorm(N_ROWS, 0, 0.85), 2),
  plate_z = round(rnorm(N_ROWS, 2.4, 0.75), 2),
  description = desc,
  delta_run_exp = round(dre, 4)
)

# Statcast missingness: spin and arm_angle are absent for a slice of rows, and a
# few pitchers lack arm_angle entirely -- the latter is what makes a whole group
# average to NaN and produces incomplete rows downstream.
out$release_spin_rate[runif(N_ROWS) < 0.02] <- NA
out$arm_angle[runif(N_ROWS) < 0.03] <- NA
out$release_extension[runif(N_ROWS) < 0.005] <- NA
dead <- sample(pit$player_name, 5)
out$arm_angle[out$player_name %in% dead] <- NA

write_csv(out, "synthetic_statcast.csv")
cat(nrow(out), "rows written\n")
print(count(out, pitch_name, sort = TRUE))
